# Achievement 4, Task 9 (Part 1)

In [1]:
#import libraries 
import numpy as np 
import pandas as pd 
import os 
import matplotlib.pyplot as plt
import seaborn as sns 
import scipy 

In [2]:
#path to instacart basket analysis 
path = r'/Users/libertyonia/Documents/Instacart Basket Analysis'

In [3]:
#import ords_prods_merge dataframe 
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_4.8.pkl'))

In [4]:
#check that prices above 100 were dropped previously
ords_prods_merge.loc[ords_prods_merge['prices'] >= 100]

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_day_of_week, order_hour_of_day, days_since_prior_order, product_id, add_to_cart_order, reordered, product_name, aisle_id, department_id, prices, _merge, price_range_loc, busiest_day, busiest_days, busiest_period_of_day, max_order, loyalty_flag, average_spending, spending_flag, median_dspo, order_behavior]
Index: []

[0 rows x 25 columns]

In [5]:
#import customers.csv
customers_df = pd.read_csv(os.path.join(path,'02 Data','Original Data', 'customers.csv'))

In [6]:
customers_df.shape

(206209, 10)

# Checking data 
#### Actions that were completed: 
- Check the column name using .head()
- Check total rows and columns 
- Check descriptive statistics 
- Check gender frequency 
- Check column datatypes 
- Check if there are any N/A aka missing values 
- Check for duplicates 

In [7]:
#check the columns in the customers dataframe 
customers_df.head(1)

user_id First Name    Surnam  Gender     STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665

In [8]:
#check the columns and rows of the customers dataframe 
#there are 206209 rows and 10 columns 
customers_df.shape

(206209, 10)

In [9]:
#check the descriptive statistics of the customers dataframe 
customers_df.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [10]:
#check the frequency by gender 
customers_df['Gender'].value_counts(dropna = False)

Male      104067
Female    102142
Name: Gender, dtype: int64

In [11]:
#check the datatypes of each column in the customers dataframe 
customers_df.dtypes

user_id          int64
First Name      object
Surnam          object
Gender          object
STATE           object
Age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

In [12]:
#check if there are an N/A values 
customers_df.isnull().sum()

user_id             0
First Name      11259
Surnam              0
Gender              0
STATE               0
Age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

In [13]:
#Check if there are any duplicates
customers_df_dups = customers_df[customers_df.duplicated()]
customers_df_dups

Empty DataFrame
Columns: [user_id, First Name, Surnam, Gender, STATE, Age, date_joined, n_dependants, fam_status, income]
Index: []

In [14]:
#Check if there are any columns with mixed datatypes 
for col in customers_df.columns.tolist():
    x = (customers_df[[col]].applymap(type) != customers_df[[col]].iloc[0].apply(type)).any(axis = 1)
if len(customers_df[x]) > 0:
    print(col)

# Question 4 & 5
Q4. Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.

Q5. Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example, check for and address missing values and duplicates, and convert any mixed-type data.

In [15]:
#Rename column Surnam to Surname 
customers_df.rename(columns = {'Surnam':'Surname'}, inplace = True)

#### Rename columns?
I renamed the Surnam column to Surname

#### Drop Columns? 
I didn't drop any columns, because I believe all columns may be relevant to future data analysis. 

#### Drop N/A values? 
There are 11,259 N/A values in the First Name column, but I decided not to drop these values since they will have unique customer ids to help differentiate similar surnames

#### Change any datatypes? Are there any mixed datatypes?
I didn't change any data types, because they were consistent with what was expected per column. Also, there are no mixed datatypes. 

#### Drop duplicates?
I didn't drop any duplicates, because there were no duplicates. 

In [16]:
customers_df.shape

(206209, 10)

# Question 6 
Combine your customer data with the rest of your prepared Instacart data. (Hint: Make sure the key columns are the same data type!)

In [17]:
#drop _merge column from the ords_prods_merge dataframe
ords_prods_merge = ords_prods_merge.drop(columns = ['_merge'])

In [18]:
#check columns in the ords_prods_merge dataframe 
ords_prods_merge.columns.tolist()

['order_id',
 'user_id',
 'eval_set',
 'order_number',
 'order_day_of_week',
 'order_hour_of_day',
 'days_since_prior_order',
 'product_id',
 'add_to_cart_order',
 'reordered',
 'product_name',
 'aisle_id',
 'department_id',
 'prices',
 'price_range_loc',
 'busiest_day',
 'busiest_days',
 'busiest_period_of_day',
 'max_order',
 'loyalty_flag',
 'average_spending',
 'spending_flag',
 'median_dspo',
 'order_behavior']

In [19]:
#check columns in customers dataframe 
customers_df.columns.tolist()

['user_id',
 'First Name',
 'Surname',
 'Gender',
 'STATE',
 'Age',
 'date_joined',
 'n_dependants',
 'fam_status',
 'income']

In [20]:
#merge ords_prods_merge dataframe with customers dataframe using 'user_id'
ords_prods_cust = ords_prods_merge.merge(customers_df, on = 'user_id', indicator = True)

In [21]:
#check value counts of _merge column
ords_prods_cust['_merge'].value_counts(dropna = False)

both          30328763
left_only            0
right_only           0
Name: _merge, dtype: int64

# Question 8 
Export this new dataframe as a pickle file so you can continue to use it in the second part of this task.

In [22]:
#check new dataframe columns 
ords_prods_cust.columns.tolist()

['order_id',
 'user_id',
 'eval_set',
 'order_number',
 'order_day_of_week',
 'order_hour_of_day',
 'days_since_prior_order',
 'product_id',
 'add_to_cart_order',
 'reordered',
 'product_name',
 'aisle_id',
 'department_id',
 'prices',
 'price_range_loc',
 'busiest_day',
 'busiest_days',
 'busiest_period_of_day',
 'max_order',
 'loyalty_flag',
 'average_spending',
 'spending_flag',
 'median_dspo',
 'order_behavior',
 'First Name',
 'Surname',
 'Gender',
 'STATE',
 'Age',
 'date_joined',
 'n_dependants',
 'fam_status',
 'income',
 '_merge']

In [23]:
#Check the columns and rows of the customers dataframe 
customers_df.shape

(206209, 10)

In [24]:
#Check the columns and rows of the ords_prods_merge dataframe 
ords_prods_merge.shape

(30328763, 24)

In [25]:
#check columns and rows of new ords_prods_cust dataframe 
ords_prods_cust.shape

(30328763, 34)

In [26]:
#Export the ords_prods_cust dataframe as a pickle 
ords_prods_cust.to_pickle(os.path.join(path,'02 Data','Prepared Data','ords_prods_cust_4-9.pkl'))